pkg/io/video/scale.go

```diff
@@ -0,0 +1,175 @@
+				rect.Max.X /= 2
+			case image.YCbCrSubsampleRatio420:
+				rect.Max.X /= 2
+				rect.Max.Y /= 2
+			}
+			return rect
+		}
+		// ycbcrRealloc reallocs image.YCbCr if needed
+		ycbcrRealloc := func(i1 *image.YCbCr) {
+			if imgScaled == nil {
+				imgScaled = image.NewYCbCr(rect, i1.SubsampleRatio)
+				imgScaledCopy = image.NewYCbCr(rect, i1.SubsampleRatio)
+			}
+			imgDst := imgScaled.(*image.YCbCr)
+			yDx := rect.Dx()
+			yDy := rect.Dy()
+			cRect := fixedRect(rect, i1.SubsampleRatio)
+			cDx := cRect.Dx()
+			cDy := cRect.Dx()
+			yLen := yDx * yDy
+			cLen := cDx * cDy
+			if len(imgDst.Y) < yLen {
+				if cap(imgDst.Y) < yLen {
+					imgDst.Y = make([]uint8, yLen)
+				}
+				imgDst.Y = imgDst.Y[:yLen]
+			}
+			if len(imgDst.Cr) < cLen {
+				if cap(imgDst.Cr) < cLen {
+					imgDst.Cr = make([]uint8, cLen)
+				}
+				imgDst.Cr = imgDst.Cr[:cLen]
+			}
+			if len(imgDst.Cb) < cLen {
+				if cap(imgDst.Cb) < cLen {
+					imgDst.Cb = make([]uint8, cLen)
+				}
+				imgDst.Cb = imgDst.Cb[:cLen]
+			}
+			*dst.y = image.Gray{Pix: imgDst.Y, Stride: imgDst.YStride, Rect: rect}
+			*dst.cb = image.Gray{Pix: imgDst.Cb, Stride: imgDst.CStride, Rect: cRect}
+			*dst.cr = image.Gray{Pix: imgDst.Cr, Stride: imgDst.CStride, Rect: cRect}
+		}
+		// ycbcrRealloc reallocs image.RGBA if needed
+		rgbaRealloc := func(i1 *image.RGBA) {
+			if imgScaled == nil {
+				imgScaled = image.NewRGBA(rect)
+				imgScaledCopy = image.NewRGBA(rect)
+			}
+			imgDst := imgScaled.(*image.RGBA)
+			l := 4 * rect.Dx() * rect.Dy()
+			if len(imgDst.Pix) < l {
+				if cap(imgDst.Pix) < l {
+					imgDst.Pix = make([]uint8, l)
+				}
+				imgDst.Pix = imgDst.Pix[:l]
+			}
+		}
+
+		return ReaderFunc(func() (image.Image, error) {
+			img, err := r.Read()
+			if err != nil {
+				return nil, err
+			}
+
+			if imgScaled == nil {
+				if height <= 0 {
+					h := img.Bounds().Dy() * width / img.Bounds().Dx()
+					rect = image.Rect(0, 0, width, h)
+				} else if width <= 0 {
+					w := img.Bounds().Dx() * height / img.Bounds().Dy()
+					rect = image.Rect(0, 0, w, height)
+				}
+				cacheScaler(rect, img.Bounds())
+			}
+
+			switch v := img.(type) {
+			case *image.RGBA:
+				rgbaRealloc(v)
+				dst := imgScaled.(*image.RGBA)
+				scalerCached.Scale(dst, rect, v, v.Rect, draw.Src, nil)
+
+				*(imgScaledCopy.(*image.RGBA)) = *dst // Clone metadata
+
+			case *image.YCbCr:
+				ycbcrRealloc(v)
+				// Scale each plane
+				*src.y = image.Gray{Pix: v.Y, Stride: v.YStride, Rect: v.Rect}
+				*src.cb = image.Gray{
+					Pix: v.Cb, Stride: v.CStride, Rect: fixedRect(v.Rect, v.SubsampleRatio),
+				}
+				*src.cr = image.Gray{
+					Pix: v.Cr, Stride: v.CStride, Rect: fixedRect(v.Rect, v.SubsampleRatio),
+				}
+				scalerCached.Scale(dst, dst.Bounds(), src, src.Bounds(), draw.Src, nil)
+
+				*(imgScaledCopy.(*image.YCbCr)) = *(imgScaled.(*image.YCbCr)) // Clone metadata
+
+			default:
+				return nil, errUnsupportedImageType
+			}
+
+			return imgScaledCopy, nil
```

Comment:
I'm curious. Why didn't you return `imgScaled` instead?

Reply:
It's necessary to set metadata each time as the next reader like ToI420 might overwrite it, but it is bit complexed since there are variety of pixel format and sub-sampling. So, I just copy the container (buffer of the slices are shared).

Reply:
Ah I see. If that's the case, then I would not declare `imgScaledCopy`, but just return in the switch case.

So, instead of,

```
*(imgScaledCopy.(*image.RGBA)) = *dst // Clone metadata
```

```
cloned := *dst // Clone metadata
return cloned, nil
```

I think it'll be as performant still, but it's more readable and simpler. What do you think?

Reply:
I think allocation degrades performance a little bit, but looks still acceptable amount.